File: src/papernestextension/controls/material/papernest_date_picker.py
from dataclasses import field
from datetime import date, datetime, timezone
from enum import Enum
import json
import re
from typing import Optional

from flet.controls.base_control import control
from flet.controls.border_radius import BorderRadiusValue
from flet.controls.control_event import ControlEventHandler, Event, EventHandler
from flet.controls.duration import DateTimeValue
from flet.controls.layout_control import LayoutControl
from flet.controls.material.form_field_control import InputBorder
from flet.controls.material.textfield import KeyboardType
from flet.controls.padding import Padding, PaddingValue
from flet.controls.text_style import TextStyle
from flet.controls.types import (
    ColorValue,
    IconDataOrControl,
    Locale,
    Number,
    StrOrControl,
    TextAlign,
)

__all__ = [
    "PaperNestDatePicker",
    "PaperNestDatePickerEntryMode",
    "PaperNestDatePickerEntryModeChangeEvent",
    "PaperNestDatePickerMode",
]


class PaperNestDatePickerMode(Enum):
    DAY = "day"
    YEAR = "year"


class PaperNestDatePickerEntryMode(Enum):
    CALENDAR = "calendar"
    INPUT = "input"
    CALENDAR_ONLY = "calendarOnly"
    INPUT_ONLY = "inputOnly"


class PaperNestDatePickerEntryModeChangeEvent(Event["PaperNestDatePicker"]):
    entry_mode: PaperNestDatePickerEntryMode


@control("PaperNestDatePicker")
class PaperNestDatePicker(LayoutControl):
    """Sélecteur de date autonome à l'apparence d'un champ Material.

    Le champ ouvre le DatePicker Material natif. Son dialogue conserve toutes
    les fonctions natives et reçoit une identité visuelle PaperNest via le
    ``builder`` Flutter de ``showDatePicker()``.
    """

    value: Optional[DateTimeValue] = None
    first_date: DateTimeValue = field(
        default_factory=lambda: datetime(year=1900, month=1, day=1)
    )
    last_date: DateTimeValue = field(
        default_factory=lambda: datetime(year=2050, month=1, day=1)
    )
    current_date: DateTimeValue = field(default_factory=datetime.now)
    locale: Optional[Locale] = field(default_factory=lambda: Locale("fr", "FR"))

    keyboard_type: KeyboardType = KeyboardType.DATETIME
    date_picker_mode: PaperNestDatePickerMode = PaperNestDatePickerMode.DAY
    entry_mode: PaperNestDatePickerEntryMode = PaperNestDatePickerEntryMode.CALENDAR

    help_text: Optional[str] = "Sélectionner une date"
    cancel_text: Optional[str] = "Annuler"
    confirm_text: Optional[str] = "Valider"
    error_format_text: Optional[str] = "Format invalide"
    error_invalid_text: Optional[str] = "Date hors limites"
    field_hint_text: Optional[str] = "jj/mm/aaaa"
    field_label_text: Optional[str] = "Date"

    switch_to_calendar_icon: Optional[IconDataOrControl] = None
    switch_to_input_icon: Optional[IconDataOrControl] = None
    barrier_color: Optional[ColorValue] = None
    inset_padding: PaddingValue = field(
        default_factory=lambda: Padding.symmetric(horizontal=16.0, vertical=24.0)
    )

    # Thème du dialogue DatePicker Material natif.
    picker_primary_color: Optional[ColorValue] = None
    picker_bgcolor: Optional[ColorValue] = None
    picker_header_bgcolor: Optional[ColorValue] = None
    picker_header_color: Optional[ColorValue] = None
    picker_border_radius: Number = 20

    autofocus: bool = False
    text_align: TextAlign = TextAlign.START

    # Apparence FormField, déclarée localement pour garder le contrôle autonome.
    label: Optional[StrOrControl] = None
    hint_text: Optional[str] = "jj/mm/aaaa"
    helper: Optional[StrOrControl] = None
    helper_text: Optional[str] = None
    helper_max_lines: Optional[int] = None
    error: Optional[StrOrControl] = None
    error_text: Optional[str] = None
    error_max_lines: Optional[int] = None
    text_size: Optional[Number] = None
    text_style: Optional[TextStyle] = None
    label_style: Optional[TextStyle] = None
    floating_label_style: Optional[TextStyle] = None
    hint_style: Optional[TextStyle] = None
    hint_max_lines: Optional[int] = None
    helper_style: Optional[TextStyle] = None
    error_style: Optional[TextStyle] = None
    prefix: Optional[StrOrControl] = None
    prefix_style: Optional[TextStyle] = None
    prefix_icon: Optional[IconDataOrControl] = None
    clear_button: bool = False
    clear_icon: Optional[IconDataOrControl] = None
    clear_tooltip: str = "Effacer"
    border: Optional[InputBorder] = None
    color: Optional[ColorValue] = None
    focused_color: Optional[ColorValue] = None
    bgcolor: Optional[ColorValue] = None
    focused_bgcolor: Optional[ColorValue] = None
    focus_color: Optional[ColorValue] = None
    border_width: Number = 1
    border_color: Optional[ColorValue] = None
    border_radius: Optional[BorderRadiusValue] = None
    focused_border_width: Optional[Number] = None
    focused_border_color: Optional[ColorValue] = None
    content_padding: Optional[PaddingValue] = None
    dense: bool = False
    collapsed: bool = False
    align_label_with_hint: bool = False
    filled: bool = False
    fill_color: Optional[ColorValue] = None
    hover_color: Optional[ColorValue] = None

    on_change: Optional[ControlEventHandler["PaperNestDatePicker"]] = None
    on_cleared: Optional[ControlEventHandler["PaperNestDatePicker"]] = None
    on_entry_mode_change: Optional[
        EventHandler[PaperNestDatePickerEntryModeChangeEvent]
    ] = None
    on_focus: Optional[ControlEventHandler["PaperNestDatePicker"]] = None
    on_blur: Optional[ControlEventHandler["PaperNestDatePicker"]] = None
    on_tap_outside: Optional[ControlEventHandler["PaperNestDatePicker"]] = None
    on_escape: Optional[ControlEventHandler["PaperNestDatePicker"]] = None

    @staticmethod
    def _date_from_event_data(event_data) -> Optional[datetime]:
        """Force une date civile pure sans décalage de fuseau."""
        if event_data is None:
            return None

        value = event_data
        if isinstance(value, datetime):
            if value.tzinfo is not None:
                value = value.astimezone()
            return datetime(value.year, value.month, value.day)
        if isinstance(value, date):
            return datetime(value.year, value.month, value.day)

        if isinstance(value, str):
            text = value.strip()
            if text.startswith("{"):
                try:
                    payload = json.loads(text)
                    value = payload.get("value") or payload.get("date") or text
                except (json.JSONDecodeError, TypeError):
                    value = text
        elif isinstance(value, dict):
            value = value.get("value") or value.get("date")

        if value is None:
            return None

        if isinstance(value, (int, float)) or (
            isinstance(value, str)
            and re.fullmatch(r"-?\d+(?:\.\d+)?", value.strip())
        ):
            timestamp = float(value)
            if abs(timestamp) > 10_000_000_000:
                timestamp /= 1000.0
            local = datetime.fromtimestamp(timestamp, tz=timezone.utc).astimezone()
            return datetime(local.year, local.month, local.day)

        if isinstance(value, str):
            text = value.strip()
            if re.match(r"^\d{4}-\d{2}-\d{2}", text):
                try:
                    return datetime.strptime(text[:10], "%Y-%m-%d")
                except ValueError:
                    return None
            for fmt in ("%d/%m/%Y", "%m/%d/%Y"):
                try:
                    parsed = datetime.strptime(text, fmt)
                    return datetime(parsed.year, parsed.month, parsed.day)
                except ValueError:
                    continue
        return None

    async def _trigger_event(self, event_name, event_data):
        if event_name == "change":
            selected_date = self._date_from_event_data(event_data)
            self.value = selected_date
        elif event_name == "cleared":
            self.value = None
        await super()._trigger_event(event_name, event_data)

    async def focus(self) -> None:
        await self._invoke_method("focus")

    async def open(self) -> None:
        await self._invoke_method("open")

    async def clear(self) -> None:
        await self._invoke_method("clear")

File: src/flutter/papernestextension/lib/src/controls/papernest_date_picker.dart
import 'package:flet/flet.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class PaperNestDatePickerControl extends StatefulWidget {
  final Control control;

  PaperNestDatePickerControl({Key? key, required this.control})
      : super(key: key ?? ValueKey("control_${control.id}"));

  @override
  State<PaperNestDatePickerControl> createState() =>
      _PaperNestDatePickerControlState();
}

class _PaperNestDatePickerControlState
    extends State<PaperNestDatePickerControl> {
  late final FocusNode _focusNode;
  DateTime? _value;
  bool _focused = false;
  bool _hovered = false;

  @override
  void initState() {
    super.initState();
    _focusNode = FocusNode();
    _focusNode.addListener(_onFocusChange);
    widget.control.addInvokeMethodListener(_invokeMethod);
    _value = _parseDate(widget.control.get("value"));
  }

  @override
  void dispose() {
    _focusNode.removeListener(_onFocusChange);
    _focusNode.dispose();
    widget.control.removeInvokeMethodListener(_invokeMethod);
    super.dispose();
  }

  void _onFocusChange() {
    if (_focused == _focusNode.hasFocus) return;
    setState(() => _focused = _focusNode.hasFocus);
    widget.control.triggerEvent(_focused ? "focus" : "blur");
  }

  Future<dynamic> _invokeMethod(String name, dynamic args) async {
    switch (name) {
      case "focus":
        _focusNode.requestFocus();
        return null;
      case "open":
        await _openPicker();
        return null;
      case "clear":
        _clearValue();
        return null;
      default:
        throw Exception("Unknown PaperNestDatePicker method: $name");
    }
  }

  DateTime? _parseDate(dynamic raw) {
    if (raw == null) return null;

    if (raw is DateTime) {
      final local = raw.isUtc ? raw.toLocal() : raw;
      return DateTime(local.year, local.month, local.day);
    }

    if (raw is num) {
      final milliseconds =
          raw.abs() > 10000000000 ? raw.toInt() : raw.toInt() * 1000;
      final local = DateTime.fromMillisecondsSinceEpoch(
        milliseconds,
        isUtc: true,
      ).toLocal();
      return DateTime(local.year, local.month, local.day);
    }

    final text = raw.toString().trim();
    if (text.isEmpty || text.toLowerCase() == "none") return null;

    final civil = RegExp(r'^(\d{4})-(\d{2})-(\d{2})').firstMatch(text);
    if (civil != null) {
      return DateTime(
        int.parse(civil.group(1)!),
        int.parse(civil.group(2)!),
        int.parse(civil.group(3)!),
      );
    }

    final parsed = DateTime.tryParse(text)?.toLocal();
    return parsed == null
        ? null
        : DateTime(parsed.year, parsed.month, parsed.day);
  }

  String _civilIso(DateTime value) =>
      '${value.year.toString().padLeft(4, '0')}-${value.month.toString().padLeft(2, '0')}-${value.day.toString().padLeft(2, '0')}';

  String _displayDate(DateTime value) =>
      '${value.day.toString().padLeft(2, '0')}/${value.month.toString().padLeft(2, '0')}/${value.year.toString().padLeft(4, '0')}';

  DatePickerEntryMode _entryMode() {
    switch (widget.control.getString("entry_mode")) {
      case "input":
        return DatePickerEntryMode.input;
      case "calendarOnly":
        return DatePickerEntryMode.calendarOnly;
      case "inputOnly":
        return DatePickerEntryMode.inputOnly;
      default:
        return DatePickerEntryMode.calendar;
    }
  }

  DatePickerMode _pickerMode() =>
      widget.control.getString("date_picker_mode") == "year"
          ? DatePickerMode.year
          : DatePickerMode.day;

  Widget _buildPickerTheme(BuildContext context, Widget? child) {
    final theme = Theme.of(context);
    final primaryColor =
        widget.control.getColor("picker_primary_color", context) ??
            const Color(0xFFF9A825);
    final backgroundColor =
        widget.control.getColor("picker_bgcolor", context) ?? Colors.white;
    final headerBackgroundColor =
        widget.control.getColor("picker_header_bgcolor", context) ??
            Colors.grey.shade900;
    final headerColor =
        widget.control.getColor("picker_header_color", context) ?? Colors.white;

    return Theme(
      data: theme.copyWith(
        colorScheme: theme.colorScheme.copyWith(
          primary: primaryColor,
          surface: backgroundColor,
        ),
        datePickerTheme: theme.datePickerTheme.copyWith(
          backgroundColor: backgroundColor,
          headerBackgroundColor: headerBackgroundColor,
          headerForegroundColor: headerColor,
          shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(
              widget.control.getDouble("picker_border_radius", 20) ?? 20,
            ),
          ),
        ),
      ),
      child: child ?? const SizedBox.shrink(),
    );
  }

  Future<void> _openPicker() async {
    if (widget.control.disabled) return;
    if (!_focusNode.hasFocus) _focusNode.requestFocus();

    final now = DateTime.now();
    final firstDate =
        _parseDate(widget.control.get("first_date")) ?? DateTime(1900, 1, 1);
    final lastDate =
        _parseDate(widget.control.get("last_date")) ?? DateTime(2050, 1, 1);
    var initialDate =
        _value ?? _parseDate(widget.control.get("current_date")) ?? now;
    if (initialDate.isBefore(firstDate)) initialDate = firstDate;
    if (initialDate.isAfter(lastDate)) initialDate = lastDate;

    final selected = await showDatePicker(
      context: context,
      initialDate: initialDate,
      firstDate: firstDate,
      lastDate: lastDate,
      currentDate: _parseDate(widget.control.get("current_date")),
      initialEntryMode: _entryMode(),
      initialDatePickerMode: _pickerMode(),
      helpText: widget.control.getString("help_text"),
      cancelText: widget.control.getString("cancel_text"),
      confirmText: widget.control.getString("confirm_text"),
      errorFormatText: widget.control.getString("error_format_text"),
      errorInvalidText: widget.control.getString("error_invalid_text"),
      fieldHintText: widget.control.getString("field_hint_text"),
      fieldLabelText: widget.control.getString("field_label_text"),
      barrierColor: widget.control.getColor("barrier_color", context),
      locale: const Locale('fr', 'FR'),
      builder: _buildPickerTheme,
      onDatePickerModeChange: (mode) {
        widget.control.triggerEvent(
          "entry_mode_change",
          mode == DatePickerEntryMode.input ? "input" : "calendar",
        );
      },
    );

    if (selected == null) return;
    final civil = DateTime(selected.year, selected.month, selected.day);
    setState(() => _value = civil);
    final iso = _civilIso(civil);
    widget.control.updateProperties({"value": iso});
    widget.control.triggerEvent("change", iso);
  }

  void _clearValue() {
    if (widget.control.disabled || _value == null) return;
    setState(() => _value = null);
    widget.control.updateProperties({"value": null});
    widget.control.triggerEvent("cleared");
    widget.control.triggerEvent("change", "");
    _focusNode.requestFocus();
  }

  InputBorder _border(BuildContext context, {required bool focused}) {
    final borderType = widget.control.getString("border") ?? "outline";
    if (borderType == "none") return InputBorder.none;

    final width = focused
        ? widget.control.getDouble("focused_border_width") ??
            widget.control.getDouble("border_width", 1)!
        : widget.control.getDouble("border_width", 1)!;
    final color = focused
        ? widget.control.getColor("focused_border_color", context) ??
            widget.control.getColor("border_color", context) ??
            Theme.of(context).colorScheme.primary
        : widget.control.getColor("border_color", context) ??
            Theme.of(context).colorScheme.outline;
    final side = width == 0
        ? BorderSide.none
        : BorderSide(color: color, width: width);

    if (borderType == "underline") {
      return UnderlineInputBorder(borderSide: side);
    }
    return OutlineInputBorder(
      borderSide: side,
      borderRadius: widget.control.getBorderRadius("border_radius") ??
          const BorderRadius.all(Radius.circular(4)),
    );
  }

  Widget _buildPrefixIcon() {
    return widget.control.buildIconOrWidget("prefix_icon") ??
        const Icon(Icons.calendar_month_outlined);
  }

  Widget? _buildClearButton() {
    final clearButton = widget.control.getBool("clear_button", false)!;
    if (!clearButton || _value == null || widget.control.disabled) return null;

    final icon = widget.control.buildIconOrWidget("clear_icon") ??
        const Icon(Icons.clear_rounded);
    return IconButton(
      icon: icon,
      tooltip: widget.control.getString("clear_tooltip", "Effacer"),
      onPressed: _clearValue,
    );
  }

  @override
  Widget build(BuildContext context) {
    final backendValue = _parseDate(widget.control.get("value"));
    if ((backendValue == null) != (_value == null) ||
        (backendValue != null &&
            _value != null &&
            _civilIso(backendValue) != _civilIso(_value!))) {
      _value = backendValue;
    }

    final theme = Theme.of(context);
    final color = widget.control.getColor("color", context);
    final focusedColor = widget.control.getColor("focused_color", context);
    final textStyle = (widget.control.getTextStyle("text_style", theme) ??
            theme.textTheme.bodyLarge ??
            const TextStyle())
        .copyWith(
      fontSize: widget.control.getDouble("text_size"),
      color: _focused ? focusedColor ?? color : color,
    );

    final bgcolor = widget.control.getColor("bgcolor", context);
    final focusedBgcolor =
        widget.control.getColor("focused_bgcolor", context);
    final fillColor = widget.control.getColor("fill_color", context) ??
        (_focused ? focusedBgcolor ?? bgcolor : bgcolor);

    final decoration = InputDecoration(
      enabled: !widget.control.disabled,
      label: widget.control.buildTextOrWidget("label"),
      labelStyle: widget.control.getTextStyle("label_style", theme),
      floatingLabelStyle:
          widget.control.getTextStyle("floating_label_style", theme),
      hintText: widget.control.getString("hint_text"),
      hintStyle: widget.control.getTextStyle("hint_style", theme),
      hintMaxLines: widget.control.getInt("hint_max_lines"),
      helper: widget.control.buildTextOrWidget("helper"),
      helperText: widget.control.getString("helper_text"),
      helperStyle: widget.control.getTextStyle("helper_style", theme),
      helperMaxLines: widget.control.getInt("helper_max_lines"),
      error: widget.control.buildTextOrWidget("error"),
      errorText: widget.control.getString("error_text"),
      errorStyle: widget.control.getTextStyle("error_style", theme),
      errorMaxLines: widget.control.getInt("error_max_lines"),
      prefix: widget.control.buildTextOrWidget("prefix"),
      prefixStyle: widget.control.getTextStyle("prefix_style", theme),
      prefixIcon: _buildPrefixIcon(),
      prefixIconConstraints:
          widget.control.getBoxConstraints("prefix_icon_constraints"),
      suffixIcon: _buildClearButton(),
      suffixIconConstraints:
          widget.control.getBoxConstraints("suffix_icon_constraints"),
      isDense: widget.control.getBool("dense", false)!,
      isCollapsed: widget.control.getBool("collapsed", false)!,
      alignLabelWithHint:
          widget.control.getBool("align_label_with_hint", false),
      filled: widget.control.getBool("filled", fillColor != null)!,
      fillColor: fillColor,
      hoverColor: widget.control.getColor("hover_color", context),
      focusColor: widget.control.getColor("focus_color", context),
      contentPadding: widget.control.getPadding("content_padding"),
      constraints: widget.control.getBoxConstraints("size_constraints"),
      border: _border(context, focused: false),
      enabledBorder: _border(context, focused: false),
      focusedBorder: _border(context, focused: true),
      disabledBorder: _border(context, focused: false),
    );

    final field = FocusableActionDetector(
      focusNode: _focusNode,
      autofocus: widget.control.getBool("autofocus", false)!,
      enabled: !widget.control.disabled,
      mouseCursor: widget.control.disabled
          ? SystemMouseCursors.basic
          : SystemMouseCursors.click,
      onShowFocusHighlight: (value) {
        if (_focused != value && mounted) setState(() => _focused = value);
      },
      onShowHoverHighlight: (value) {
        if (_hovered != value && mounted) setState(() => _hovered = value);
      },
      shortcuts: const <ShortcutActivator, Intent>{
        SingleActivator(LogicalKeyboardKey.enter): ActivateIntent(),
        SingleActivator(LogicalKeyboardKey.space): ActivateIntent(),
      },
      actions: <Type, Action<Intent>>{
        ActivateIntent: CallbackAction<ActivateIntent>(
          onInvoke: (_) {
            _openPicker();
            return null;
          },
        ),
      },
      child: Material(
        type: MaterialType.transparency,
        child: InkWell(
          mouseCursor: widget.control.disabled
              ? SystemMouseCursors.basic
              : SystemMouseCursors.click,
          canRequestFocus: false,
          onTap: widget.control.disabled ? null : _openPicker,
          hoverColor: widget.control.getColor("hover_color", context),
          focusColor: widget.control.getColor("focus_color", context),
          borderRadius: widget.control.getBorderRadius("border_radius"),
          child: InputDecorator(
            isFocused: _focused,
            isHovering: _hovered,
            isEmpty: _value == null,
            decoration: decoration,
            child: Text(
              _value == null ? "" : _displayDate(_value!),
              style: textStyle,
              textAlign:
                  widget.control.getTextAlign("text_align", TextAlign.start),
            ),
          ),
        ),
      ),
    );

    return LayoutControl(control: widget.control, child: field);
  }
}
